### PR TITLE
feat(cli): add `--rules` CLI flag

### LINKED ISSUE #283: feat(cli): add `--rules` to display the total number of registered rules
Ideally, the command `oxc lint --rules` should display

```
Correctness (number of rules):
- A
- B
Nursery (number of rules):
- C
- D
Total: number of rules
```

### PR DESCRIPTION
**Pull request by shannonrothe**
Closes #283

### CODE REVIEW

File: crates/oxc_cli/src/lint/mod.rs [hunk 1]
```diff
@@ -179,4 +181,10 @@ mod test {
             get_lint_options("lint --ignore-pattern ./test --ignore-pattern bar.js foo.js");
         assert_eq!(options.ignore_pattern, vec![String::from("./test"), String::from("bar.js")]);
     }
+
+    #[test]
+    fn list_rules_true() {
+        let options = get_lint_options("lint foo.js --rules");
```
Comment (Boshen) [inline crates/oxc_cli/src/lint/mod.rs:187]: It would be great if this works without the file arg as well

File: crates/oxc_linter/src/rule.rs [hunk 2]
```diff
@@ -53,3 +53,13 @@ impl RuleCategory {
         }
     }
 }
+
+impl AsRef<str> for RuleCategory {
```
Comment (Boshen) [inline crates/oxc_linter/src/rule.rs:57]: In Rust, we normall implement the Display trait

File: crates/oxc_cli/src/lint/runner.rs [hunk 3]
```diff
@@ -115,6 +130,18 @@ impl LintRunner {
         }
     }
 
+    fn dump_rules(&self) {
+        let mut stdout = BufWriter::new(std::io::stdout());
+        for (category, rules) in &self.registered_rules {
```
Comment (Boshen) [inline crates/oxc_cli/src/lint/runner.rs:135]: We should sort the names somewhere, or is it already sorted?

File: crates/oxc_cli/src/lint/runner.rs [hunk 4]
```diff
@@ -29,6 +29,8 @@ pub struct LintRunner {
     options: LintOptions,
 
     linter: Arc<Linter>,
+
+    registered_rules: FxHashMap<RuleCategory, Vec<RuleEnum>>,
```
Comment (Boshen) [inline crates/oxc_cli/src/lint/runner.rs:33]: Does it make sense to get the rules without storing it here? We have the `RULES` global

File: crates/oxc_cli/src/lint/runner.rs (hunk 3 above)
Comment (shannonrothe) [reply crates/oxc_cli/src/lint/runner.rs:135]: How should they be sorted? Could we just have a static const vec that holds our desired order and sort here before we iterate?

File: crates/oxc_cli/src/lint/mod.rs (hunk 1 above)
Comment (shannonrothe) [reply crates/oxc_cli/src/lint/mod.rs:187]: The path is required though 🤔

File: crates/oxc_cli/src/lint/runner.rs (hunk 4 above)
Comment (shannonrothe) [reply crates/oxc_cli/src/lint/runner.rs]: Oh sorry, I misinterpreted the  requirements here, we just want to output all available rules right (and their associated counts) + the total? Should be fixed [here](https://github.com/Boshen/oxc/pull/290/commits/a947bd06c7c97fc8ffe7b3e698f21452675a70cb)

File: crates/oxc_cli/src/lint/mod.rs (hunk 1 above)
Comment (shannonrothe) [reply crates/oxc_cli/src/lint/mod.rs:187]: I'm not super familiar with `clap` but is it possible to make an arg optional depending on another arg? i.e. `PATH` can be optional if `--rules` is present

Comment (shannonrothe) [reply crates/oxc_cli/src/lint/mod.rs]: Seems like we want this: https://docs.rs/clap/4.2.2/clap/struct.Arg.html#method.required_unless_present 👍🏼

File: crates/oxc_cli/src/lint/runner.rs (hunk 3 above)
Comment (Boshen) [reply crates/oxc_cli/src/lint/runner.rs]: We should be able to clone the names here and then sort them.

Comment (Boshen) [reply crates/oxc_cli/src/lint/runner.rs]: You can also utilize `BTreeSet`

File: crates/oxc_cli/src/lint/runner.rs [hunk 5]
```diff
@@ -36,6 +36,22 @@ pub struct LintRunner {
 #[diagnostic(help("{0:?} seems like a minified file"))]
 pub struct MinifiedFileError(pub PathBuf);
 
+fn print_rules() {
```
[the remark this answers is not in the document]
Comment (shannonrothe) [reply crates/oxc_cli/src/lint/runner.rs:39]: That's what I had originally 😂

File: crates/oxc_cli/src/lint/runner.rs (hunk 3 above)
Comment (shannonrothe) [reply crates/oxc_cli/src/lint/runner.rs]: It seems they're already in the order defined by the enum. Is that OK?

**Review by Boshen — APPROVED** [review]

**Comment by Boshen** [comment]
Nice work! Thank you!

**Pull request merged**